File: Backend/app/schemas/fx.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class FXRateItem(BaseModel):
    base: str = Field(min_length=3, max_length=3)
    quote: str = Field(min_length=3, max_length=3)
    rate: float = Field(gt=0)
    timestamp: datetime
    observed_on: date
    source: str
    is_synthetic: bool = False


class FXRatesResponse(BaseModel):
    data: list[FXRateItem]
    timestamp: datetime


class FXHistoryPoint(BaseModel):
    date: date
    rate: float = Field(gt=0)


class FXHistoryStats(BaseModel):
    min: float = Field(gt=0)
    max: float = Field(gt=0)
    avg: float = Field(gt=0)
    volatility: float = Field(ge=0, le=1)
    standard_deviation: float = Field(ge=0)


class FXHistoryResponse(BaseModel):
    pair: str
    base: str = Field(min_length=3, max_length=3)
    quote: str = Field(min_length=3, max_length=3)
    period: Literal["1d", "7d", "30d", "90d", "1y"]
    data: list[FXHistoryPoint]
    stats: FXHistoryStats
    data_points: int = Field(ge=1)
    real_data_points: int = Field(ge=0)
    synthetic_data_points: int = Field(ge=0)
    contains_synthetic: bool = False
    source: str


class FXCandleItem(BaseModel):
    timestamp: datetime
    open: float = Field(gt=0)
    high: float = Field(gt=0)
    low: float = Field(gt=0)
    close: float = Field(gt=0)
    source: str


class FXCandlesStats(BaseModel):
    min: float = Field(gt=0)
    max: float = Field(gt=0)
    avg: float = Field(gt=0)
    volatility: float = Field(ge=0, le=1)
    standard_deviation: float = Field(ge=0)
    change_percent: float


class FXCandlesResponse(BaseModel):
    pair: str
    base: str = Field(min_length=3, max_length=3)
    quote: str = Field(min_length=3, max_length=3)
    range: Literal["1d", "7d", "30d", "90d", "1y"]
    interval: Literal["1min", "5min", "15min", "30min", "1h", "4h", "1day", "1week"]
    data: list[FXCandleItem]
    stats: FXCandlesStats
    data_points: int = Field(ge=1)
    source: str

File: frontend/src/lib/types/fx.ts
import { z } from 'zod';
import { CurrencyCodeSchema } from './currency';

/**
 * Historical FX rate data point
 */
export const FXHistoryPointSchema = z.object({
  date: z.string(),
  rate: z.number().positive(),
  high: z.number().positive().optional(),
  low: z.number().positive().optional(),
  open: z.number().positive().optional(),
  close: z.number().positive().optional(),
});

export type FXHistoryPoint = z.infer<typeof FXHistoryPointSchema>;

/**
 * FX History response schema
 */
export const FXHistoryResponseSchema = z.object({
  pair: z.string(),
  base: CurrencyCodeSchema,
  quote: CurrencyCodeSchema,
  period: z.enum(['1d', '7d', '30d', '90d', '1y']),
  data: z.array(FXHistoryPointSchema),
  statistics: z.object({
    min: z.number().positive(),
    max: z.number().positive(),
    average: z.number().positive(),
    volatility: z.number().min(0).max(1),
    standardDeviation: z.number().nonnegative(),
  }),
  dataPoints: z.number().int().positive(),
  realDataPoints: z.number().int().nonnegative(),
  syntheticDataPoints: z.number().int().nonnegative(),
  containsSynthetic: z.boolean(),
  source: z.string(),
});

export type FXHistoryResponse = z.infer<typeof FXHistoryResponseSchema>;

/**
 * FX Statistics for a currency pair
 */
export const FXStatisticsSchema = z.object({
  base: CurrencyCodeSchema,
  quote: CurrencyCodeSchema,
  currentRate: z.number().positive(),
  averageRate: z.number().positive(),
  minRate: z.number().positive(),
  maxRate: z.number().positive(),
  volatility: z.number().min(0).max(1),
  trend: z.enum(['up', 'down', 'stable']),
  changePercent: z.number(),
});

export type FXStatistics = z.infer<typeof FXStatisticsSchema>;

/**
 * Volatility level enum
 */
export type VolatilityLevel = 'low' | 'medium' | 'high';

/**
 * Get volatility level from numeric value
 */
export const getVolatilityLevel = (volatility: number): VolatilityLevel => {
  if (volatility <= 0.3) return 'low';
  if (volatility <= 0.7) return 'medium';
  return 'high';
};

/**
 * FX conversion result
 */
export const FXConversionResultSchema = z.object({
  fromAmount: z.number().positive(),
  toAmount: z.number().positive(),
  fromCurrency: CurrencyCodeSchema,
  toCurrency: CurrencyCodeSchema,
  rate: z.number().positive(),
  timestamp: z.string().datetime(),
});

export type FXConversionResult = z.infer<typeof FXConversionResultSchema>;

/**
 * Best rate comparison data
 */
export interface BestRateComparisonData {
  currentRate: number;
  bestRate: number;
  bestRateDate: string;
  amountAtCurrentRate: number;
  amountAtBestRate: number;
  difference: number;
  percentageDifference: number;
}

/**
 * Savings estimate data
 */
export interface SavingsEstimate {
  amount: number;
  currentRate: number;
  averageRate: number;
  currentConversion: number;
  averageConversion: number;
  savings: number;
  savingsPercentage: number;
  isPositiveSavings: boolean;
}
